### ISSUE TITLE
backfill the `status`, `route`, and `ref` columns of the `exchanges` table

### ISSUE THREAD

**Issue by whit537** [gratipay/gratipay.com#2779]
Right now the `exchanges.status` column has NULL values. It would be great to constrain that column to non-NULL, but it would take some effort with the Balanced API to determine the proper value for previous exchanges. [IRC](https://botbot.me/freenode/gratipay/2014-09-23/?msg=22175953&page=2)
##

**Comment by Changaco** [comment]
Added to the [Balanced shutdown](https://github.com/gratipay/gratipay.com/milestones/Balanced%20shutdown) milestone.

**Comment by whit537** [comment]
We have a user getting bitten by #3302, which we decided to fix with this ticket.

https://gratipay.freshdesk.com/helpdesk/tickets/1867

**Comment by whit537** [comment]
I'm working on manually fixing this for the 1867 case, and am blocking payday (#3307) on it.

**Comment by whit537** [comment]
[IRC](https://botbot.me/freenode/gratipay/2015-04-02/?msg=35700392&page=2)

**Comment by whit537** [comment]
I believe the `destination` parameter here identifies the route for the most recently failing deposit:

https://dashboard.balancedpayments.com/#/marketplaces/MP12Xw5lL6iaILtqImIoroDL/events/EVfe85f7b4d70211e4b84a0663eb3e293e

However, the same parameter on the previous nine deposits (some successful, some not) is not present in `exchange_routes.address`:

https://dashboard.balancedpayments.com/#/marketplaces/MP12Xw5lL6iaILtqImIoroDL/events/EV3c627cc8d15011e496a0065bec47279b
https://dashboard.balancedpayments.com/#/marketplaces/MP12Xw5lL6iaILtqImIoroDL/events/EV9610ff60c92111e4b8dc02da29bdd175
https://dashboard.balancedpayments.com/#/marketplaces/MP12Xw5lL6iaILtqImIoroDL/events/EV864e9340c69111e4a8dc02da29bdd175
https://dashboard.balancedpayments.com/#/marketplaces/MP12Xw5lL6iaILtqImIoroDL/events/EV58831afac2bb11e49819065bec47279b
https://dashboard.balancedpayments.com/#/marketplaces/MP12Xw5lL6iaILtqImIoroDL/events/EV2f43cbb2b89811e4a512065bec47279b
https://dashboard.balancedpayments.com/#/marketplaces/MP12Xw5lL6iaILtqImIoroDL/events/EV501740b0b31811e4adbd065bec47279b
https://dashboard.balancedpayments.com/#/marketplaces/MP12Xw5lL6iaILtqImIoroDL/events/EV4decb9d4ae3911e4962c02da29bdd175
https://dashboard.balancedpayments.com/#/marketplaces/MP12Xw5lL6iaILtqImIoroDL/events/EV54ec1920a82e11e4ad9702da29bdd175
https://dashboard.balancedpayments.com/#/marketplaces/MP12Xw5lL6iaILtqImIoroDL/events/EV54966e04a29711e4972f065bec47279b

**Comment by whit537** [comment]
We are expecting one of the old deposits to fail, at which time Balanced will post back so we can update `exchanges.status`. We'll need to insert the old `exchange_route` before we can connect the exchange to it. For now I'm only worrying about the most recent failure.

**Comment by whit537** [comment]
I set `exchanges.route` for the exchange in question and clicked "Replay" on the Balanced dashboard:

https://dashboard.balancedpayments.com/#/marketplaces/MP12Xw5lL6iaILtqImIoroDL/events/EVfe85f7b4d70211e4b84a0663eb3e293e

It's still in the "retrying" state, and I'm not seeing the status updated on the [user's History page](https://gratipay.com/sudoroom/history/).

**Comment by whit537** [comment]
Here's the error in Sentry (cf. #3302):

https://app.getsentry.com/gratipay/gratipay-com/group/60437761/

**Comment by whit537** [comment]
I'm not getting good visibility into the status of the webhook postback. From Sentry it looks like we haven't generated any errors since I hit "Replay," but Sentry samples (right? how?), so not every request is recorded. If the postback works then there also won't be a Sentry entry. The absence of a Sentry entry either means it worked or it didn't but Sentry isn't showing us the specific failure instance.

All we get in the Balanced dashboard is this:

![screen shot 2015-04-02 at 5 08 53 pm](https://cloud.githubusercontent.com/assets/134455/6973755/fc28ce2e-d95a-11e4-8ca6-7cefbf27b248.png)

**Comment by whit537** [comment]
Using the Chrome inspector I've found the json endpoint for callbacks, which includes this:

```
      "attempts_remaining": 9,
      "attempts": 1,
      "attempt_limit": 10,
```

**Comment by whit537** [comment]
> the callback will be retried up to 10 times with an initial retry delay of 10 minutes which will increase exponentially with each failure for the next 7 days.

https://docs.balancedpayments.com/1.1/api/callbacks/

**Comment by whit537** [comment]
Weeeeeeeelllllll, I clicked Replay more than 10 minutes ago, and it's still reporting as retrying with 9 attempts left.

**Comment by whit537** [comment]
So I could update the user's balance manually, but I don't see that I can cancel the callback. :-(

**Comment by whit537** [comment]
If I run `record_exchange_result` manually and then start payday, and the callback succeeds while payday is running, then we will send the user twice as much money as we should. That'd be a mess to clean up.

**Comment by whit537** [comment]
But at this point we're down to the wire on payday.

**Comment by Changaco** [comment]
> then we will send the user twice as much money as we should

No, we won't.

**Comment by Changaco** [comment]
Of course you also have to update the status of the exchange, not just the balance.

**Comment by whit537** [comment]
```python
from gratipay.main import Participant
from gratipay.billing.exchanges import record_an_exchange

sudoroom = Participant.from_username('sudoroom')
record_an_exchange( Participant.db
                  , 67813
                  , 'failed'
                  , 'Funds unavailable due to action by RDFI or other legal action'
                  , sudoroom
                   )
```

**Comment by whit537** [comment]
[IRC](https://botbot.me/freenode/gratipay/2015-04-02/?msg=35707187&page=2)

**Comment by whit537** [comment]
```
[gratipay] $ heroku config -s | honcho run -e /dev/stdin ./env/bin/python unblock.py
Traceback (most recent call last):
  File "unblock.py", line 12, in <module>
    , foo
  File "/Users/whit537/personal/gratipay/gratipay.com/gratipay/billing/exchanges.py", line 318, in record_exchange_result
    assert route
AssertionError
[gratipay] $
```

**Comment by whit537** [comment]
I clicked "Replay" and still get the "retring" message.

**Comment by whit537** [comment]
Succeeded! :-)

**Comment by whit537** [comment]
Okay!

So as part of this ticket we'll need to look for any further failed webhooks in Balanced and replay them.

**Comment by rohitpaulk** [comment]
For PayPal, 

```sql
-- Successful PayPal transfers
SELECT COUNT(*) 
  FROM exchanges 
 WHERE amount < 0 
   AND note LIKE '%PayPal%';
--> 4156

-- Reversed PayPal transfers
SELECT COUNT(*)
  FROM exchanges 
 WHERE amount > 0 
   AND note LIKE '%PayPal%';
--> 4

-- Successful PayPal transfers without a status
SELECT COUNT(*) 
  FROM exchanges 
 WHERE amount < 0
   AND status IS NULL
   AND note LIKE '%PayPal%';
--> 1727

-- Reversed PayPal transfers without a status
SELECT COUNT(*)
  FROM exchanges 
 WHERE amount > 0
   AND status IS NULL
   AND note LIKE '%PayPal%';
--> 4
```

**Comment by rohitpaulk** [comment]
https://botbot.me/freenode/gratipay/2014-09-23/?msg=22175953&page=2 isn't going to be easy, we'll need to find the reversed exchange for each of the ones that failed, and delete that. I'm tempted to just mark all of them as 'succeeded', and focus on just the routes instead.

**Comment by whit537** [comment]
If we're making the API calls anyway it might make sense to do #2443 together with this one.

**Comment by whit537** [comment]
See https://github.com/gratipay/inside.gratipay.com/issues/323#issuecomment-135588541 for another instance of the above one-off fix.

**Comment by whit537** [comment]
Over at https://github.com/gratipay/inside.gratipay.com/issues/308#issuecomment-143008427 I made a download of the JSON for [all of our Balanced transactions](https://github.com/gratipay/logs/tree/master/misc/308/raw/balanced/transactions). Might be a discrepancy but we're close. Use the scripts [over there](https://github.com/gratipay/logs/tree/master/misc/308/raw/balanced) as a basis for backfilling routes here .

**Comment by whit537** [comment]
> Do all of these necessarily have to be present in Balanced?

No.

> Could they also be old exchanges from Stripe or so?

Yes.

**Comment by whit537** [comment]
We're at 77,360 `exchanges`, [still](https://github.com/gratipay/gratipay.com/issues/2779#issuecomment-57045390) with 37,557 with NULL status.

Here's what a record looks like in our database:

```sql
=# select * from exchanges where id=[n];
┌─────┬───────────────────────────────┬────────┬──────┬─────────────┬──────────┬──────┬────────┬───────┐
│ id  │           timestamp           │ amount │ fee  │ participant │ recorder │ note │ status │ route │
├─────┼───────────────────────────────┼────────┼──────┼─────────────┼──────────┼──────┼────────┼───────┤
│ [n] │ 2012-07-27 20:51:27.828555+00 │   9.32 │ 0.68 │ [username]  │ ¤        │ ¤    │ ¤      │     ¤ │
└─────┴───────────────────────────────┴────────┴──────┴─────────────┴──────────┴──────┴────────┴───────┘
(1 row)
```

And here's what we want to match it up with from `transactions/*.json`:

```python
(Pdb) pp thing
{u'amount': 1000,
 u'appears_on_statement_as': u'BAL*www.gittip.com',
 u'created_at': u'2012-07-27T20:51:27.505700Z',
 u'currency': u'USD',
 u'description': u'[username]',
 u'failure_reason': None,
 u'failure_reason_code': None,
 u'href': u'/debits/[]',
 u'id': u'[]',
 u'links': {u'card_hold': u'[]',
            u'customer': u'[]',
            u'dispute': None,
            u'order': None,
            u'source': u'[]'},
 u'meta': {},
 u'status': u'succeeded',
 u'transaction_number': u'[]',
 u'updated_at': u'2012-07-27T20:51:27.532539Z'}
(Pdb)
```

**Comment by whit537** [comment]
Start on this, got a hot mess of a script atm:

```python
#!/usr/bin/env python -u
from __future__ import absolute_import, division, print_function, unicode_literals

import os, json
from gratipay import wireup
from decimal import Decimal as D

db = wireup.db(wireup.env())
dirname = os.path.join(os.path.dirname(__file__), 'transactions')
i = 0
I = 500000
class Done(Exception): pass

with db.get_cursor() as cur:
    try:
        for filename in reversed(sorted(os.listdir(dirname))):
            data = json.load(open(os.path.join(dirname, filename)))

            for key in data:
                if key in ('links', 'meta'): continue
                things = reversed(data[key])

                if key == 'card_holds':
                    continue  # We don't track these.

                if key != 'debits':
                    continue  # Let's start with credit card charges.

                for thing in things:
                    route = cur.one( "SELECT * FROM exchange_routes WHERE address='/cards/'||%s"
                                   , (thing['links']['source'],)
                                    )
                    if route is None:
                        import pdb; pdb.set_trace()
                        continue
                    username = cur.one( "SELECT username FROM participants WHERE id=%s"
                                      , (route.participant,)
                                       )

                    params = dict( status   = thing['status']
                                 , amount   = D(thing['amount']) / 100
                                 , ts       = thing['created_at']
                                 , route_id = route.id
                                 , username = username
                                  )
                    mogrified = cur.mogrify("""\
                    WITH rows AS (
                        UPDATE exchanges
                           SET status = %(status)s
                             , route = %(route_id)s
                         WHERE participant = %(username)s
                           AND amount + fee = %(amount)s
                           AND ((   (%(ts)s::timestamptz - "timestamp") < interval '0'
                                AND (%(ts)s::timestamptz - "timestamp") > interval '-1s'
                                ) OR (
                                    (%(ts)s::timestamptz - "timestamp") > interval '0'
                                AND (%(ts)s::timestamptz - "timestamp") < interval '1s'
                                ))
                     RETURNING participant
                    ) SELECT * FROM rows;
                    """, params)
                    res = cur.one(mogrified)
                    print(params['ts'], res, username)

                    if i > I:
                        raise Done()
                    i += 1
    except Done:
        pass
    print()
```

**Comment by whit537** [comment]
| `grep` | `wc -l` |
|:-------|----------:|
| `": 1\.$"` | 8,145 |
| `"Need to insert route for"`| 11,079 |
| `"Need to link"` | 8 |
| `"ExchangeMismatch: "` | 28,078 |
| (`"ExchangeMismatch: .*got None"` | 28,078) |
| `"TooMany"` | 526 |
| Sum | 47,836 |
| `"Z | "` | 47,573 |


```bash
$ grep "raise" backfill.log | sort | uniq
    raise ExchangeMismatch(nupdated)
    raise TooMany(self.rowcount, lo, hi)
$
```

```python
#!/usr/bin/env python -u
from __future__ import absolute_import, division, print_function, unicode_literals

import json, os, sys, traceback
from gratipay import wireup
from decimal import Decimal as D
from pprint import pprint

db = wireup.db(wireup.env())
dirname = os.path.join(os.path.dirname(__file__), 'transactions')
i = 0
I = float('inf')
class Done(Exception): pass

class ExchangeMismatch(Exception):
    def __str__(self):
        return "Expected 1 update, got {}.".format(self.args[0])


def process_thing(cur, thing):
    customer = cur.one( "SELECT participants.*::participants "
                        "FROM participants "
                        "WHERE balanced_customer_href='/customers/'||%s"
                      , (thing['links']['customer'],)
                       )
    route = cur.one( "SELECT * FROM exchange_routes WHERE address='/cards/'||%s"
                   , (thing['links']['source'],)
                    )

    if customer is None and route is None:
        print("Need to link {} to ... {}(?) and create a route for {}."
              .format( thing['links']['customer']
                     , thing['description']
                     , thing['links']['source']
                      ))
        return
    elif participant is None:
        print("Need to link {} to ... {}?"
              .format(thing['links']['customer'], thing['description']))
        return
    elif route is None:
        print("Need to insert route for {} and {}."
              .format(participant.username, thing['links']['source']))
        return

    # Use the username that is current as of this transaction.
    username = cur.one( "SELECT username FROM participants WHERE id=%s"
                      , (route.participant,)
                       )

    params = dict( status   = thing['status']
                 , amount   = D(thing['amount']) / 100
                 , ts       = thing['created_at']
                 , route_id = route.id
                 , username = username
                  )
    mogrified = cur.mogrify("""\
    WITH rows AS (
        UPDATE exchanges
           SET status = %(status)s
             , route = %(route_id)s
         WHERE participant = %(username)s
           AND amount + fee = %(amount)s
           AND ((   (%(ts)s::timestamptz - "timestamp") < interval '0'
                AND (%(ts)s::timestamptz - "timestamp") > interval '-1s'
                ) OR (
                    (%(ts)s::timestamptz - "timestamp") > interval '0'
                AND (%(ts)s::timestamptz - "timestamp") < interval '1s'
                ))
     RETURNING 1 AS count
    ) SELECT * FROM rows;
    """, params)
    print("Updating exchange(s) for {}: ".format(username), end='')
    nupdated = cur.one(mogrified)
    if nupdated != 1:
        raise ExchangeMismatch(nupdated)
    print("{}.".format(nupdated))


with db.get_cursor() as cur:
    try:
        for filename in reversed(sorted(os.listdir(dirname))):
            data = json.load(open(os.path.join(dirname, filename)))

            for key in data:
                if key in ('links', 'meta'): continue
                things = reversed(data[key])

                if key == 'card_holds':
                    continue  # We don't track these.

                if key != 'debits':
                    continue  # Let's start with credit card charges.

                for thing in things:
                    print("{} | ".format(thing['created_at']), end='')

                    try:
                        process_thing(cur, thing)
                    except KeyboardInterrupt:
                        raise
                    except:
                        print("Exception! Thing:")
                        pprint(thing)
                        print('-'*78)
                        traceback.print_exc(file=sys.stdout)
                        print('='*78)

                    if i > I: raise Done()
                    i += 1
    except Done:
        pass
    print()
```

**Comment by whit537** [comment]
Alright, we've got duplicate exchange routes ((`network`, `address`) is not unique), and we've also got duplicates in `participants.balanced_customer_href`.

**Comment by rohitpaulk** [comment]
Whoops :)

**Comment by rohitpaulk** [comment]
I'm working on https://github.com/gratipay/gratipay.com/pull/3356. 

> Before we start filling the route values for previous exchanges, we should make sure that no new exchanges are created with a NULL route. As of now, record-an-exchange.spt is the only place where this happens (MassPay uses it, so that too). That's what this PR aims to fix.

**Comment by whit537** [comment]
!m @rohitpaulk :dancer:

**Comment by whit537** [comment]
Some of both kinds of duplicates are due to merge/takeover. Not sure if *all*, but that seems likely.

**Comment by whit537** [comment]
Spot-checking the `None` matches suggests that a 1-second time interval is too tight to match. We should loosen that to 2 seconds and see if we overmatch any (we also constrain by username and amount, so I wouldn't expect so).

**Comment by whit537** [comment]
This one is off by almost thirty seconds.

**Comment by whit537** [comment]
We have a bunch with non-null status, but null route:

 ```sql
=> select count(*) from exchanges where route is null and status is not null;
┌───────┐
│ count │
├───────┤
│ 31620 │
└───────┘
(1 row)
```

**Comment by whit537** [comment]
I upped the threshold to one minute, and I'm not seeing any `NoExchanges` except for a handful of special cases where we charged people manually, out-of-band from payday. It looks like if we up the threshold 15 minutes we should be able to pick these off. Again, that shouldn't introduce duplicates because we only charge once per week.

**Comment by whit537** [comment]
Alright, so I guess we need to resolve duplicate routes and customers before proceeding here. Reticketed as #3806.

**Comment by whit537** [comment]
```sh
#!/bin/sh
patterns=(": 1\.$" 
          "No route for" 
          "NoExchange" 
          "TooMany")
for pattern in "${patterns[@]}"; do
    printf "%-14s %5s\n" "\"$pattern\"" `grep "$pattern" backfill.log | wc -l`
done
```

```bash
$ balanced/tally-backfill.sh
": 1\.$"       33918
"No route for" 11186
"NoExchange"    2308
"TooMany"        322
$
```

**Comment by whit537** [comment]
We want 47,573 of the first, and zero of the rest.

I believe most of the `NoExchange` are in fact *already linked* exchanges, from after when we started recording status and route.

**Comment by whit537** [comment]
- [ ] deduplicate routes and customers—#3806
 - [ ] backfill `participants.balanced_customer_href`—we're missing a few, I think?
 - [ ] create missing routes—definitely missing a bunch of these
 - [ ] modify the `exchanges` schema to store a URL and a JSON payload

**Comment by whit537** [comment]
Haven't even touched refunds, or bank deposits or reversals yet. We don't actually need those in order to be able to issue refunds against Balanced this week, which is our primary objective atm.

**Comment by whit537** [comment]
Though conceptually it makes sense to get this done for *all* transaction types while we're here.

**Comment by whit537** [comment]
Still seeing one `NoExchange`: a special case where the manual exchange was recorded wrong, with a second exchange to correct. The `amount` of the first exchange is out relative to Balanced.

**Comment by whit537** [comment]
Okay, cool. We're down to one `NoExchange`, with two others picked off by expanding the time window to ±15m, and the rest falling under `Already linked`.

```
": 1\.$"            33920
"Already linked"     2305
"NoExchange"            1
"AmbiguousExchange"     0
"No route for"      11186
"TooMany"             322
```

**Comment by whit537** [comment]
I think we can backfill missing customers and routes as part of this script, once both are deduplicated.

**Comment by whit537** [comment]
Got the numbers to line up:

```
": 1\.$"              33920
"Already linked"       2305
"NoExchange: "            1
"AmbiguousExchange: "     0
"No route for"        11186
"raise TooMany"         161
---------------------------
                      47573
```

```sh
#!/bin/sh
patterns=(": 1\.$" 
          "Already linked"
          "NoExchange: " 
          "AmbiguousExchange: "
          "No route for" 
          "raise TooMany")
N=0
for pattern in "${patterns[@]}"; do
    n=`grep "$pattern" backfill.log | wc -l`
    N=`expr $n + $N`
    printf "%-21s %5s\n" "\"$pattern\"" $n
done
echo "---------------------------"
printf "%-21s %5s\n" "" $N
```

**Comment by whit537** [comment]
"raise TooMany" should go to zero with #3806, and we'll handle "No route for" by creating routes on the fly while looping.

**Comment by whit537** [comment]
```python
#!/usr/bin/env python -u
from __future__ import absolute_import, division, print_function, unicode_literals

import json, os, sys, traceback
from gratipay import wireup
from decimal import Decimal as D
from pprint import pprint

db = wireup.db(wireup.env())
dirname = os.path.join(os.path.dirname(__file__), 'transactions')
i = 0
I = float('inf')
class Done(Exception): pass

class ExchangeMismatch(Exception):
    def __str__(self):
        return "{} {} {}\n\n{}".format(*self.args)
class NoExchange(ExchangeMismatch): pass
class AmbiguousExchange(ExchangeMismatch): pass


def process_thing(cur, thing):
    print('{} -> '.format(thing['description']), end='')

    participants = cur.all( "SELECT username FROM participants "
                            "WHERE balanced_customer_href='/customers/'||%s"
                          , (thing['links']['customer'],)
                           )
    print('{} | '.format(', '.join(participants)), end='')

    route = cur.one( "SELECT * FROM exchange_routes WHERE address='/cards/'||%s"
                   , (thing['links']['source'],)
                    )
    if route is None:
        if not participants:
            print("No route for {}.".format(thing['links']['source']))
        else:
            print("No route for {}.".format(thing['links']['source']))
        return

    # Use the username that is current as of this db transaction.
    username = cur.one( "SELECT username FROM participants WHERE id=%s"
                      , (route.participant,)
                       )

    params = dict( status   = thing['status']
                 , amount   = D(thing['amount']) / 100
                 , ts       = thing['created_at']
                 , route_id = route.id
                 , username = username
                  )

    nalready = cur.one("""\
        SELECT count(*) FROM exchanges
         WHERE status = %(status)s
           AND route = %(route)s
           AND participant = %(username)s
           AND amount + fee = %(amount)s
           AND ((   (%(ts)s::timestamptz - "timestamp") < interval '0'
                AND (%(ts)s::timestamptz - "timestamp") > interval '-15m'
                ) OR (
                    (%(ts)s::timestamptz - "timestamp") > interval '0'
                AND (%(ts)s::timestamptz - "timestamp") < interval '15m'
                ))
    """, dict(params, route=route.id))
    if nalready == 1:
        print("Already linked!")
        return

    mogrified = cur.mogrify("""\
    WITH rows AS (
        UPDATE exchanges
           SET status = %(status)s
             , route = %(route_id)s
         WHERE (status is NULL or status = %(status)s)
           AND route is NULL
           AND participant = %(username)s
           AND amount + fee = %(amount)s
           AND ((   (%(ts)s::timestamptz - "timestamp") < interval '0'
                AND (%(ts)s::timestamptz - "timestamp") > interval '-15m'
                ) OR (
                    (%(ts)s::timestamptz - "timestamp") > interval '0'
                AND (%(ts)s::timestamptz - "timestamp") < interval '15m'
                ))
     RETURNING 1 AS count
    ) SELECT * FROM rows;
    """, params)
    print("Updating exchanges for {}: ".format(username), end='')
    nupdated = cur.one(mogrified, default=0)
    Error = None
    if nupdated == 0:
        Error = NoExchange
    elif nupdated > 1:
        Error = AmbiguousExchange
    if Error:
        raise Error(params['username'], params['amount'], params['ts'], mogrified)
    print("{}.".format(nupdated))


with db.get_cursor() as cur:
    try:
        for filename in reversed(sorted(os.listdir(dirname))):
            data = json.load(open(os.path.join(dirname, filename)))

            for key in data:
                if key in ('links', 'meta'): continue
                things = reversed(data[key])

                if key == 'card_holds':
                    continue  # We don't track these.

                if key != 'debits':
                    continue  # Let's start with credit card charges.

                for thing in things:
                    print("{} | ".format(thing['created_at']), end='')

                    try:
                        process_thing(cur, thing)
                    except KeyboardInterrupt:
                        raise
                    except:
                        print("Exception! Thing:")
                        pprint(thing)
                        print('-'*78)
                        traceback.print_exc(file=sys.stdout)
                        print('='*78)

                    if i > I: raise Done()
                    i += 1
    except Done:
        pass
    print()
    raise Exception  # trigger rollback
```

**Comment by whit537** [comment]
Once this is done we can `drop column exchanges.participant`, since we'll have a reference through `exchange_routes`. That'll knock one out for #835.

**Comment by whit537** [comment]
Alright, so after some poking on #3806, I'm now thinking that we should pick one of the duplicates rather than deduplicating.

**Comment by whit537** [comment]
Working on this at Blue Canary with @kaguillera. Caught him up to speed, and now diving back in. We just discovered that later transactions in Balanced have an `exchange_id` and a `participant_id`. Looks like it starts around August, 2014. Let's see how much ambiguity that removes ...

**Comment by whit537** [comment]
We just verified that every transaction with a `exchange_id` also has a `participant_id`.

**Comment by whit537** [comment]
GitHub comments as version control. :cat: 
```python
#!/usr/bin/env python -u
from __future__ import absolute_import, division, print_function, unicode_literals

import json, os, sys, traceback
from gratipay import wireup
from decimal import Decimal as D
from pprint import pprint
from gratipay.models.exchange_route import ExchangeRoute
from gratipay.models.participant import Participant

db = wireup.db(wireup.env())
dirname = os.path.join(os.path.dirname(__file__), 'transactions')
i = 0
I = float('inf')
class Done(Exception): pass

class ExchangeMismatch(Exception):
    def __str__(self):
        return "{} {} {}\n\n{}".format(*self.args)
class NoExchange(ExchangeMismatch): pass
class AmbiguousExchange(ExchangeMismatch): pass


def get_participant(cur, customer_id, old_username):
    participant = cur.one("SELECT participants.*::participants FROM participants "
                          "WHERE balanced_customer_href='/customers/'||%s", (customer_id,))
    if participant is None:
        participant = cur.one("SELECT participants.*::participants FROM participants "
                              "WHERE username=%s", (old_username,))
        #XXX Check for username changes!
    if participant is None:
        raise Exception("Unable to get a participant.")
    return participant


def create_route(cur, thing):
    participant = get_participant(cur, thing['links']['customer'], thing['description'])
    route = ExchangeRoute.insert(participant, 'balanced-cc', '/cards/'+thing['links']['source'])
    return route


def pick_route(cur, routes, thing):
    import pdb; pdb.set_trace()
    route = None
    return route


def process_thing(cur, thing):
    print(thing['description'], end='')
    route, participant = None, None

    # Try to pick a route.
    routes = cur.all( "SELECT * FROM exchange_routes "
                      "WHERE network='balanced-cc' and address='/cards/'||%s"
                    , (thing['links']['source'],)
                     )
    if not routes:
        print("No route for {}.".format(thing['links']['source']), end='')
        route = create_route(cur, thing)
    elif len(routes) > 1:
        print("{} routes for {}.".format(len(routes), thing['links']['source'], end=''))
        route = pick_route(cur, routes, thing)
    else:
        route = routes[0]

    # Ack.
    if type(route.participant) is long:
        route.set_attributes(participant=Participant.from_id(route.participant))

    # Bail if we don't have a route.
    if not route:
        print()
        return

    # Use the username that is current as of this db transaction.
    username = cur.one( "SELECT username FROM participants WHERE id=%s"
                      , (route.participant,)
                       )

    params = dict( status   = thing['status']
                 , amount   = D(thing['amount']) / 100
                 , ts       = thing['created_at']
                 , route_id = route.id
                 , username = username
                  )

    nalready = cur.one("""\
        SELECT id FROM exchanges
         WHERE status = %(status)s
           AND route = %(route)s
           AND participant = %(username)s
           AND amount + fee = %(amount)s
           AND ((   (%(ts)s::timestamptz - "timestamp") < interval '0'
                AND (%(ts)s::timestamptz - "timestamp") > interval '-15m'
                ) OR (
                    (%(ts)s::timestamptz - "timestamp") > interval '0'
                AND (%(ts)s::timestamptz - "timestamp") < interval '15m'
                ))
    """, dict(params, route=route.id))
    if nalready == 1:
        print("Already linked!")
        return

    mogrified = cur.mogrify("""\
    WITH rows AS (
        UPDATE exchanges
           SET status = %(status)s
             , route = %(route_id)s
         WHERE (status is NULL or status = %(status)s)
           AND route is NULL
           AND participant = %(username)s
           AND amount + fee = %(amount)s
           AND ((   (%(ts)s::timestamptz - "timestamp") < interval '0'
                AND (%(ts)s::timestamptz - "timestamp") > interval '-15m'
                ) OR (
                    (%(ts)s::timestamptz - "timestamp") > interval '0'
                AND (%(ts)s::timestamptz - "timestamp") < interval '15m'
                ))
     RETURNING 1 AS count
    ) SELECT * FROM rows;
    """, params)
    print("Updating exchanges for {}: ".format(username), end='')
    nupdated = cur.one(mogrified, default=0)
    Error = None
    if nupdated == 0:
        Error = NoExchange
    elif nupdated > 1:
        Error = AmbiguousExchange
    if Error:
        raise Error(params['username'], params['amount'], params['ts'], mogrified)
    print("{}.".format(nupdated))


with db.get_cursor() as cur:
    try:
        for filename in reversed(sorted(os.listdir(dirname))):
            data = json.load(open(os.path.join(dirname, filename)))

            for key in data:
                if key in ('links', 'meta'): continue
                things = reversed(data[key])

                if key == 'card_holds':
                    continue  # We don't track these.

                if key != 'debits':
                    continue  # Let's start with credit card charges.

                for thing in things:
                    print("{} | ".format(thing['created_at']), end='')

                    try:
                        process_thing(cur, thing)
                    except KeyboardInterrupt:
                        raise
                    except:
                        print("Exception! Thing:")
                        pprint(thing)
                        print('-'*78)
                        traceback.print_exc(file=sys.stdout)
                        print('='*78)

                    if i > I: raise Done()
                    i += 1
    except Done:
        pass
    print()
    raise Exception  # trigger rollback
```

**Comment by rohitpaulk** [comment]
@whit537 - https://gist.github.com/ :stuck_out_tongue:

**Comment by whit537** [comment]
Yeah, yeah. :P

**Comment by whit537** [comment]
@rohitpaulk et al. Moving to a PR, #3807.

**Comment by whit537** [comment]
Got this far enough for the Balanced Shutdown, taking off that milestone now.

**Comment by whit537** [comment]
Done for the three TransferWise exchanges to date.

**Comment by whit537** [comment]
Moving to a new PR for the remaining exchanges: #3912.
